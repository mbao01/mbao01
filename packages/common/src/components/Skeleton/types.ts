export type Size = 2 | 4 | 8 | 12 | 16 | 24 | 32 | 48 | 64 | "full";

export type SkeletonProps = React.HTMLAttributes<HTMLDivElement> & {
  width?: Size;
  height?: Size;
  round?: boolean;
  className?: string;
};
