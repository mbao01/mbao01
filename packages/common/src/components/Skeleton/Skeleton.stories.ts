import type { Meta, StoryObj } from "@storybook/react";

import { Skeleton } from "./Skeleton";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Skeleton,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Skeleton>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    width: 32,
  },
};

export const RoundSkeleton: Story = {
  args: {
    round: true,
    width: 16,
    height: 16,
  },
};

export const LargeSkeleton: Story = {
  args: {
    width: 32,
    height: 32,
  },
};
