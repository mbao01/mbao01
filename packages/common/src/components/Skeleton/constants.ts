import { cva } from "../../helpers";

export const getSkeletonClasses = cva("skeleton", {
  variants: {
    variant: { pulse: "animate-pulse" },
    width: {
      2: "w-2",
      4: "w-4",
      8: "w-8",
      12: "w-12",
      16: "w-16",
      24: "w-24",
      32: "w-32",
      48: "w-48",
      64: "w-64",
      full: "w-full",
    },
    height: {
      2: "h-2",
      4: "h-4",
      8: "h-8",
      12: "h-12",
      16: "h-16",
      24: "h-24",
      32: "h-32",
      48: "h-48",
      64: "h-64",
      full: "h-full",
    },
    round: {
      true: "rounded-full",
      false: "rounded-md",
    },
  },
  defaultVariants: {
    width: "full",
    height: 4,
    round: false,
  },
});
