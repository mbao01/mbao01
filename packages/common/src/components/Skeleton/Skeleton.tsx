import c from "clsx";
import { type SkeletonProps } from "./types";
import { getSkeletonSize } from "./constants";

export const Skeleton = ({
  className,
  round,
  width,
  height,
  ...props
}: SkeletonProps) => {
  return (
    <div
      {...props}
      className={c(
        "skeleton",
        getSkeletonSize({ width, height }),
        {
          "rounded-md": !round,
          "rounded-full": round,
        },
        className
      )}
    />
  );
};
