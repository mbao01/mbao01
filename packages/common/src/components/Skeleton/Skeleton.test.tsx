import { render, screen } from "@testing-library/react";
import { Skeleton } from "./Skeleton";

describe("Skeleton", () => {
  it("renders a skeleton", () => {
    const { asFragment } = render(<Skeleton data-testid="skeleton" />);

    expect(screen.getByTestId("skeleton")).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("has a skeleton with pulse animation", () => {
    const { asFragment } = render(<Skeleton variant="pulse" data-testid="skeleton" />);

    expect(screen.getByTestId("skeleton")).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a round skeleton", () => {
    const { asFragment } = render(<Skeleton data-testid="skeleton" round />);
    const skeletonEl = screen.getByTestId("skeleton");

    expect(skeletonEl).toBeVisible();
    expect(skeletonEl).toHaveClass("rounded-full");
    expect(asFragment()).toMatchSnapshot();
  });

  it.each([2, 64, "full"] as const)("renders a square skeleton (%s/sq)", (size) => {
    const { asFragment } = render(
      <Skeleton data-testid="skeleton" round width={size} height={size} />
    );
    const skeletonEl = screen.getByTestId("skeleton");

    expect(skeletonEl).toBeVisible();
    expect(skeletonEl).toHaveClass(`w-${size} h-${size}`);
    expect(asFragment()).toMatchSnapshot();
  });
});
