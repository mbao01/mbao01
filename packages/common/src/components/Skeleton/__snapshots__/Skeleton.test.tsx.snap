// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Skeleton > renders a round skeleton 1`] = `
<DocumentFragment>
  <div
    class="skeleton w-full h-4 rounded-full"
    data-testid="skeleton"
  />
</DocumentFragment>
`;

exports[`Skeleton > renders a skeleton 1`] = `
<DocumentFragment>
  <div
    class="skeleton w-full h-4 rounded-md"
    data-testid="skeleton"
  />
</DocumentFragment>
`;

exports[`Skeleton > renders a square skeleton (2/sq) 1`] = `
<DocumentFragment>
  <div
    class="skeleton w-2 h-2 rounded-full"
    data-testid="skeleton"
  />
</DocumentFragment>
`;

exports[`Skeleton > renders a square skeleton (64/sq) 1`] = `
<DocumentFragment>
  <div
    class="skeleton w-64 h-64 rounded-full"
    data-testid="skeleton"
  />
</DocumentFragment>
`;

exports[`Skeleton > renders a square skeleton (full/sq) 1`] = `
<DocumentFragment>
  <div
    class="skeleton w-full h-full rounded-full"
    data-testid="skeleton"
  />
</DocumentFragment>
`;
